notion-server/src/ethereum/mod.rs

```diff
@@ -226,36 +230,62 @@ impl Simulation {
 
     /// Creates a new shard block and returns the
     /// index of the created shard block
-    pub fn create_shard_block(&mut self, args: args::CreateShardBlock) -> Result<u32> {
-        if let Some(shard_chain) = self.shard_chains.get_mut(args.shard_chain_index as usize) {
-            let shard_block = ShardBlock::try_from(args.shard_block)?;
-
-            // TODO: Run each transaction (which will update the EE state for that shard)
-            // Questions to answer:
-            //   * What if the decoding of the base64 data string fails? Remove this transaction from the block?  Send back error value as result?
-            //   * What if executing the EE code fails with the given data? (Same options as above?)
-            // Example code from previous brainstorm:
-            //        let transactions = shard_block.transactions
-            //
-            //        for transaction in shard_block.transactions {
-            //            // This executes everything and presumably also updates the EE State on the shard
-            //            let ee = transaction.execution_environment;
-            //            let input_data = transaction.data;
-            //
-            //            let code = self.beacon_chain.get(ee);
-            //            let runtime = RootRuntime::new(&code, shard_ee_state_or_something_similar);
-            //            runtime.execute(input_data);
-            //        }
-
-            shard_chain.shard_blocks.push(shard_block);
-            Ok((shard_chain.shard_blocks.len() - 1) as u32)
-        } else {
-            Err(Error::OutOfBounds {
+    fn create_shard_block(&mut self, args: args::CreateShardBlock) -> Result<u32> {
+        let mut shard_chain = self
+            .shard_chains
+            .get_mut(args.shard_chain_index as usize)
+            .context(OutOfBounds {
                 message: format!("No shard chain exists at index: {}", args.shard_chain_index),
-            })
+            })?;
```

Comment:
Thinking over this a little more, it might be beneficial to rewrite the `OutOfBounds` error like this:

```rust
// Name to be bike shedded.
#[derive(Debug)]
pub enum Thing {
    ExecutionEnvironment,
    ShardChain,
}

#[derive(Snafu)]
pub enum Error {
    #[snafu(display("No {:?} exists at index: {}", thing, index)]
    OutOfBounds {
        index: usize,
        thing: Thing,
    }
}
```

This takes more advantage of `snafu` and should make the errors more consistent.

Reply:
I don't yet feel comfortable enough with `snafu` to implement that myself, would it be okay to keep it for now?